src/main/java/org/candlepin/controller/ConsumerManager.java

```diff
@@ -69,4 +92,99 @@ public void updateLastCheckIn(Consumer consumer) {
             eventAdapter.publish(cloudCheckInEvent);
         }
     }
+
+    /**
+     * Sets the {@link Environment}s of the {@link Consumer}s that correspond to the provided consumer UUIDs.
+     * Consumers will be unchanged if they already exist in the provided environments and are in the right
+     * priority. Other consumers will be removed from their existing environments and set to the provided
+     * environments. The ordering of the provided environment IDs dictates the priority. The first environment
+     * ID in the list being the top priority and the last environment ID in the list being the least priority.
+     * All updated consumers will have their content access certificate revoked.
+     *
+     * @param consumerUuids
+     *  the UUIDs of the consumers that should have the environments set
+     *
+     * @param targetEnvIds
+     *  the IDs of the {@link Environment}s that should be set for the consumers
+     *
+     * @param owner
+     *  the owner of the consumers
+     *
+     * @throws IllegalArgumentException
+     *  if the owner is null
+     *
+     * @throws BadRequestException
+     *  if the owner is not in SCA content access mode, if one or more of provided consumer UUIDs or
+     *  environment IDs is unknown to the owner, or if the target environments contains null or a duplicate
+     *  value.
+     *
+     * @return
+     *  the UUIDs of the {@link Consumer}s that were updated and had their environments set
+     */
+    public Set<String> setConsumersEnvironments(List<String> consumerUuids, List<String> targetEnvIds,
+        Owner owner) {
+
+        if (owner == null) {
+            throw new IllegalArgumentException("owner is null");
+        }
+
+        if (consumerUuids == null || consumerUuids.isEmpty() || targetEnvIds == null ||
+            targetEnvIds.isEmpty()) {
+            return new HashSet<>();
+        }
+
+        if (!ContentAccessMode.ORG_ENVIRONMENT.toDatabaseValue().equals(owner.getContentAccessMode())) {
+            throw new BadRequestException(i18n.tr("Owner is not in SCA content access mode"));
+        }
+
+        Set<String> unknownConsumerUuids = consumerCurator
+            .getNonExistentConsumerUuids(consumerUuids, owner.getKey());
+        if (!unknownConsumerUuids.isEmpty()) {
+            throw new BadRequestException(i18n.tr("Unkown consumer UUIDs: {0}",
+                unknownConsumerUuids));
+        }
+
+        Set<String> unknownEnvironmentIds = envCurator
+            .getNonExistentEnvironmentIds(targetEnvIds, owner);
+        if (!unknownEnvironmentIds.isEmpty()) {
+            throw new BadRequestException(i18n.tr("Unknown environment IDs: {0}",
```

Comment:
The service/controller level should not be throwing http- or user-facing exceptions. These are things that should be validated ahead of time and then handed over in a known good state.

Reply:
Ah yeah! Moved this validation to the Resource layer